unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/StepsController/ActionController.cs

```diff
@@ -3,151 +3,187 @@
 using System.Collections;
 using System.Collections.Generic;
 using UnityEngine;
-using static BuildModeAction;
+using static BuildInWorldCompleteAction;
 
 public class ActionController : MonoBehaviour
 {
-    public ActionListview actionListview;
 
     public BuilderInWorldEntityHandler builderInWorldEntityHandler;
 
     public System.Action OnUndo, OnRedo;
 
 
-    List<BuildModeAction> actionsMade = new List<BuildModeAction>();
+    List<BuildInWorldCompleteAction> actionsMade = new List<BuildInWorldCompleteAction>();
 
-    int currentStepIndex = 0;
+    int currentUndoStepIndex = 0;
+    int currentRedoStepIndex = 0;
 
-    private void Awake()
-    {
-        if(actionListview != null)
-            actionListview.OnActionSelected += GoToAction;
-    }
-
-    public void ClearActionList()
+    public void ResetActionList()
     {
         actionsMade.Clear();
+        currentUndoStepIndex = 0;
+        currentRedoStepIndex = 0;
     }
 
-    public void GoToAction(BuildModeAction action)
+    public void GoToAction(BuildInWorldCompleteAction action)
     {
         int index = actionsMade.IndexOf(action);
-        int stepsAmount = currentStepIndex - index;
+        int stepsAmount = currentUndoStepIndex - index;
 
         for(int i = 0; i <= Mathf.Abs(stepsAmount); i++)
         {
             if (stepsAmount > 0)
             {
                 UndoCurrentAction();
-                if (currentStepIndex > 0)
-                    currentStepIndex--;
+                if (currentUndoStepIndex > 0)
+                    currentUndoStepIndex--;
             }
             else
             {
                 RedoCurrentAction();
-                if (currentStepIndex + 1 < actionsMade.Count)
-                    currentStepIndex++;
+                if (currentUndoStepIndex + 1 < actionsMade.Count)
+                    currentUndoStepIndex++;
             }
         }
 
     }
 
     public void TryToRedoAction()
     {
-        if (currentStepIndex < actionsMade.Count)
-        {
-            if (currentStepIndex + 1 < actionsMade.Count)
-            {
-                if(currentStepIndex == 0 && actionsMade[currentStepIndex].isDone)
-                    currentStepIndex++;
-                else if(currentStepIndex != 0)
-                    currentStepIndex++;
-                RedoCurrentAction();
-            }
-        }
+        if (currentRedoStepIndex >= actionsMade.Count || currentRedoStepIndex < 0) return;
+
+        RedoCurrentAction();
+
+        if (currentRedoStepIndex + 1 < actionsMade.Count)
+            currentRedoStepIndex++;
+
+        if (currentUndoStepIndex < actionsMade.Count - 1) currentUndoStepIndex++;
+
+        Debug.Log("Redo:  Current actions " + actionsMade.Count + "   Current undo index " + currentUndoStepIndex + "   Current redo index " + currentRedoStepIndex);
```

Comment:
Remember to put those `Debug.Log` under a static `VERBOSE` bool as we do in the rest of the code.